clusterloader2/pkg/imagepreload/imagepreload.go

```diff
@@ -0,0 +1,161 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package imagepreload
+
+import (
+	"strings"
+	"time"
+
+	"github.com/spf13/pflag"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+	"k8s.io/perf-tests/clusterloader2/pkg/config"
+	"k8s.io/perf-tests/clusterloader2/pkg/framework"
+	"k8s.io/perf-tests/clusterloader2/pkg/framework/client"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement/util/informer"
+)
+
+const (
+	informerTimeout = time.Minute
+	manifest        = "$GOPATH/src/k8s.io/perf-tests/clusterloader2/pkg/imagepreload/manifests/daemonset.yaml"
+	namespace       = "preload"
+	pollingInterval = 5 * time.Second
+	pollingTimeout  = 5 * time.Minute
+)
+
+var images = pflag.StringSlice("node-preload-images", []string{}, "List of images to preload on each node in the test cluster before executing tests")
+
+type controller struct {
+	config          *config.ClusterLoaderConfig
+	framework       *framework.Framework
+	templateMapping map[string]interface{}
+	images          []string
+}
+
+// Setup ensures every node in cluster preloads given list of images before starting tests.
+// It does it by creating a Daemonset that call "docker pull" and awaits for Node object to be updated.
+// As a side-effect of the image preloading, size of Node objects is increased.
+//
+// Preloading is skipped in kubemark or if no images have been specified.
+func Setup(conf *config.ClusterLoaderConfig, f *framework.Framework) error {
+	mapping, err := config.GetMapping(conf)
+	if err != nil {
+		return err
+	}
+
+	ctl := &controller{
+		config:          conf,
+		framework:       f,
+		templateMapping: mapping,
+		images:          *images,
+	}
+	return ctl.PreloadImages()
+}
+
+func (i *controller) PreloadImages() error {
+	if len(*images) == 0 {
+		klog.Warning("No images specified. Skipping image preloading")
+		return nil
+	}
+	if i.config.ClusterConfig.Provider == "kubemark" {
+		klog.Warning("Image preloading is disabled in kubemark")
```

Comment:
Oh, this is unfortunate as it will make kubemark tests to differ even more than they do today.
I think we haven't thought of that...
@wojtek-t, any thoughts?

Reply:
I think we should solve it directly in kubemark.
Basically, how I would imagine this is to:
- on startup of hollow-kubelet, it list images from the node it's running on (it requires passing additional flags to it so that it can do that)
- it creates fakeDocker client with with that list of images
- and uses it for its whole lifetime

but that requires handling separate in kubemark - shouldn't be part of this PR.

Reply:
I think we should file an issue for it anyway - this will be useful independently from this PR.